central/policycategory/service/service_impl.go

```diff
@@ -0,0 +1,136 @@
+	resp := new(v1.GetPolicyCategoriesResponse)
+
+	parsedQuery, err := search.ParseQuery(query.GetQuery(), search.MatchAllIfEmpty())
+	if err != nil {
+		return nil, errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+	categories, err := s.policyCategoriesDatastore.SearchRawPolicyCategories(ctx, parsedQuery)
+	if err != nil {
+		return nil, err
+	}
+	resp.Categories = s.convertCategoriesToV1Categories(categories)
+
+	return resp, nil
+}
+
+func (s *serviceImpl) PostPolicyCategory(ctx context.Context, request *v1.PostPolicyCategoryRequest) (*v1.PolicyCategory, error) {
+	if request.GetPolicyCategory() == nil {
+		return nil, errors.New("empty request, policy category not specified")
+	}
+	if !validateName.MatchString(request.GetPolicyCategory().GetName()) {
+		return nil, errors.New(invalidNameErrString)
+	}
+	category, err := s.policyCategoriesDatastore.AddPolicyCategory(ctx, ToStorageProto(request.GetPolicyCategory()))
+	if err != nil {
+		return nil, err
+	}
+	return ToV1Proto(category), nil
+}
+
+func (s *serviceImpl) RenamePolicyCategory(ctx context.Context, request *v1.NewRenamePolicyCategoryRequest) (*v1.Empty, error) {
+	if !validateName.MatchString(request.GetNewCategoryName()) {
+		return nil, errors.New(invalidNameErrString)
+	}
+	return &v1.Empty{}, s.policyCategoriesDatastore.RenamePolicyCategory(ctx, request.GetId(), request.GetNewCategoryName())
+}
+
+func (s *serviceImpl) DeletePolicyCategory(ctx context.Context, request *v1.NewDeletePolicyCategoryRequest) (*v1.Empty, error) {
+	return &v1.Empty{}, s.policyCategoriesDatastore.DeletePolicyCategory(ctx, request.GetId())
+
+}
+
+// RegisterServiceServer registers this service with the given gRPC Server.
+func (s *serviceImpl) RegisterServiceServer(grpcServer *grpc.Server) {
+	v1.RegisterPolicyCategoryServiceServer(grpcServer, s)
+}
+
+// RegisterServiceHandler registers this service with the given gRPC Gateway endpoint.
+func (s *serviceImpl) RegisterServiceHandler(ctx context.Context, mux *runtime.ServeMux, conn *grpc.ClientConn) error {
+	return v1.RegisterPolicyServiceHandler(ctx, mux, conn)
+}
+
+// AuthFuncOverride specifies the auth criteria for this API.
+func (s *serviceImpl) AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error) {
+	return ctx, authorizer.Authorized(ctx, fullMethodName)
+}
+
+func (s *serviceImpl) getPolicyCategory(ctx context.Context, id string) (*v1.PolicyCategory, error) {
+	if id == "" {
+		return nil, errors.Wrap(errox.InvalidArgs, "Policy category ID must be provided")
+	}
+	category, exists, err := s.policyCategoriesDatastore.GetPolicyCategory(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.Wrapf(errox.NotFound, "policy with ID '%s' does not exist", id)
+	}
+	return ToV1Proto(category), nil
+
+}
+
+func (s *serviceImpl) convertCategoriesToV1Categories(categories []*storage.PolicyCategory) []*v1.PolicyCategory {
+	v1Categories := make([]*v1.PolicyCategory, len(categories))
```

Comment:
this should be 0, len(categories). Otherwise, this has len(categories) nil values in the front